### PR TITLE
Programmable pool functional tests

### PR DESCRIPTION
**Pull request by altonen**
This PR adds functional tests for programmable pools and implements an ECTL-type solution familiar from the Qtum whitepaper (the implementation of the ECTL replacement is done only after the testnet).

### CODE REVIEW

File: test/functional/assets/pooltester.json
Comment (muursh) [inline test/functional/assets/pooltester.json:71]: return type is null?

Comment (altonen) [reply test/functional/assets/pooltester.json:71]: I think it means that the function doesn't return anything

Comment (muursh) [reply test/functional/assets/pooltester.json:71]: Alrighty. So more `void` than `NULL`. Doesn't python use `None` for that rather than `null`?

Comment (muursh) [reply test/functional/assets/pooltester.json:71]: Ignore me. I've just noticed this is json not python. I'd skipped to the next file without noticing due to GH's stupid design

File: pallets/utxo/src/lib.rs
Comment (iljakuklic) [inline pallets/utxo/src/lib.rs:1095]: Not for this PR and not a major issue either. In general I don't like having `bool` as a function parameter since it is not obvious what it is for at the call site. Maybe a wee custom enum like `enum FundTx { Yes, No }` would make things cleaner? It is a bit more involved to use it though.

File: pallets/pp/src/lib.rs
Comment (muursh) [inline pallets/pp/src/lib.rs:147]: This is a bit nitpicky but conscrit is an annoying "word"

Comment (altonen) [reply pallets/pp/src/lib.rs:147]: I'll think of something else

Comment (muursh) [reply pallets/pp/src/lib.rs:147]: It's not all that important to me. I'm happy to move forwards with this name even if it annoys me. I haven't had chance to do anything but a brief skim so far. I'll look through your pr properly later

Comment (altonen) [reply pallets/pp/src/lib.rs:147]: Truth be told, I don't like and I think Amin doesn't like it either but `send_consensus_critical_p2pk` was too verbose in my opinion

Comment (muursh) [reply pallets/pp/src/lib.rs:147]: Yeah I agree it's a bit long. Let's leave it for now. If someone has a good idea before this is approved let's change it if not then I guess we deal with it. If nobody has a better idea there's not much we can do

File: pallets/pp/src/lib.rs
Comment (muursh) [inline pallets/pp/src/lib.rs:286]: An awful lot of acc_ids here.

**Comment by muursh** [comment]
6.3KB for the largest test contract... Do we have any ideas for the size of a "normal" contract?

**Comment by muursh** [comment]
Also the CI has run twice and failed in different ways each time...

File: pallets/utxo/src/lib.rs
Comment (altonen) [reply pallets/utxo/src/lib.rs:698]: If this appears on your screen it means you're doing something malicious so I don't think you're entitled to a better error message.

How would you reword this though?

Comment (muursh) [reply pallets/utxo/src/lib.rs:698]: Honestly not sure. I'll have a think in the morning

File: pallets/utxo/src/lib.rs
Comment (altonen) [reply pallets/utxo/src/lib.rs:694]: Very proud :D, you have to find creative ways to entertain yourself when implementing smart contract TX validation

File: pallets/pp/src/lib.rs
Comment (altonen) [reply pallets/pp/src/lib.rs:286]: I'll come up with a better name for the first variable

File: pallets/pp/src/lib.rs
Comment (altonen) [reply pallets/pp/src/lib.rs:313]: I haven't. IIRC, `ChainExtension` provides  a way to charge gas but I'll have to look into it

File: pallets/utxo/src/lib.rs
Comment (altonen) [reply pallets/utxo/src/lib.rs:461]: This `weight` variable is really badly named and actually tells how much gas the call can use. This was just a temporary hack to get something working, next PR will fix this issue and take the gas fee from the `CallPP` UTXO.

Comment (muursh) [reply pallets/utxo/src/lib.rs:461]: So it's not weight in the substrate sense of weight?

**Comment by altonen** [comment]
I can't comment on the contract sizes, maybe we could try porting some famous Solidity smart contract to ink! and see how large they are.

As for the test failures, are you able to get it fail consistently on your local machine? If I reduce the block time to 6 seconds and ran the tests, they work almost every time. I haven't tried running them multiple times with 1min block time because they take so long. Anyway, the error message has actually nothing to do with pools but there is a finalization error reported by Grandpa which shuts down one of the nodes and then the tests fail.

File: pallets/utxo/src/lib.rs
Comment (altonen) [reply pallets/utxo/src/lib.rs:461]: I'm not certain of the exact words Substrate uses but weight tells how much gas the `CallPP` can use. Maybe it's the same thing. Regardless, this is going to be addressed when we implement proper gas fees.

**Comment by muursh** [comment]
@altonen I have other issues locally 
`error: internal compiler error: encountered incremental compilation error with evaluate_obligation(fff6d7aa47675a40-bf6b8c131a5d36f6)
  |
  = help: This is a known issue with the compiler. Run `cargo clean -p mintlayer_core` or `cargo clean` to allow your project to compile
  = note: Please follow the instructions below to create a bug report with the provided information
  = note: See <https://github.com/rust-lang/rust/issues/84970> for more information

thread 'rustc' panicked at 'Found unstable fingerprints for evaluate_obligation(fff6d7aa47675a40-bf6b8c131a5d36f6): Ok(EvaluatedToOkModuloRegions)', /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/compiler/rustc_query_system/src/query/plumbing.rs:644:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md`

**Comment by muursh** [comment]
>As for the test failures, are you able to get it fail consistently on your local machine?

As discussed earlier no. I can run them locally under normal conditions with a 1min/block rate with no issues. The only time I have an issue is if I force them to run in a resource constrained env.

**Comment by altonen** [comment]
@iljakuklic just informed me that the bug (something about chain not being finalized) is reproducible with `node-template` so I think this is a Parity issue.

**Comment by iljakuklic** [comment]
> @iljakuklic just informed me that the bug (something about chain not being finalized) is reproducible with `node-template` so I think this is a Parity issue.

Looks very similar to an issue [1] which has been since closed. I am able to reproduce it with `node-template` semi-reliably.

[1] https://github.com/paritytech/substrate/issues/6727

**Comment by muursh** [comment]
> I am able to reproduce it with node-template semi-reliably

An up to date node-template? What commit?

**Comment by iljakuklic** [comment]
> > I am able to reproduce it with node-template semi-reliably
> 
> An up to date node-template? What commit?

I used `1ea0883`, will try on master tomorrow.

File: pallets/pp/src/lib.rs
Comment (altonen) [reply pallets/pp/src/lib.rs]: I renamed them to `submit_c2pk_tx()` and `submit_c2c_tx()`. The consensus-criticalness is implied

**Comment by altonen** [comment]
This time all tests also passed

File: pallets/pp/src/lib.rs
Comment (altonen) [reply pallets/pp/src/lib.rs:313]: Substrate provides `charge_weight()` which we can use to charge gas from the caller of the `ChainExtension`. I think it makes sense to do that in the PR that actually implements gas fees.

**Review by zorvan — APPROVED** [review]

File: pallets/utxo/src/lib.rs
Comment (muursh) [reply pallets/utxo/src/lib.rs:698]: That's a far better way of phrasing it

File: test/functional/feature_smart_contract_test.py
Comment (muursh) [inline test/functional/feature_smart_contract_test.py:135]: What is this data?

Comment (muursh) [reply test/functional/feature_smart_contract_test.py:135]: I see it everywhere

**Review by muursh — APPROVED** [review]

File: test/functional/feature_smart_contract_test.py
Comment (altonen) [reply test/functional/feature_smart_contract_test.py:135]: Selector (like function ID) for the constructor so the Wasm VM knows which function to call when it's instantiating the smart contract. It was a hard lesson to learn that you must always provide a selector..

Comment (zorvan) [reply test/functional/feature_smart_contract_test.py:135]: These are Selectors pointing to pooltester.json file (i.e "selector": "0xed4b9d1b")

**Pull request merged**
